pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/LimitingHeadOffsetReader.java

```diff
@@ -52,51 +51,35 @@ public LimitingHeadOffsetReader(
         CacheBuilder.newBuilder()
             .ticker(ticker)
             .expireAfterWrite(1, TimeUnit.MINUTES)
-            .build(
-                new CacheLoader<String, PslSourceOffset>() {
-                  @Override
-                  public PslSourceOffset load(String key) throws Exception {
-                    assert CACHE_KEY.equals(key);
+            .build(CacheLoader.from(this::loadHeadOffset));
+  }
 
-                    Map<Partition, Offset> partitionOffsetMap = new HashMap<>();
-                    for (int i = 0; i < topicPartitionCount; i++) {
-                      try {
-                        partitionOffsetMap.put(
-                            Partition.of(i),
-                            Offset.of(
-                                topicStatsClient
-                                    .computeHeadCursor(topic, Partition.of(i))
-                                    .get()
-                                    .getHeadCursor()
-                                    .getOffset()));
-                      } catch (Throwable t) {
-                        throw new IllegalStateException(
-                            String.format(
-                                "Unable to compute head cursor for topic partition: [%s,%d]",
-                                topic, i),
-                            t);
-                      }
-                    }
-                    return PslSourceOffset.builder().partitionOffsetMap(partitionOffsetMap).build();
-                  }
-                });
+  private Offset loadHeadOffset(Partition partition) {
+    try {
+      return Offset.of(topicStatsClient.computeHeadCursor(topic, partition).get().getOffset());
+    } catch (Throwable t) {
+      throw new IllegalStateException(
+          String.format(
+              "Unable to compute head cursor for topic partition: [%s,%d]",
+              topic, partition.value()),
+          t);
+    }
   }
 
   @Override
   public PslSourceOffset getHeadOffset() {
     try {
-      return cachedHeadOffsets.get(CACHE_KEY);
+      Map<Partition, Offset> partitionOffsetMap = new HashMap<>();
+      for (int i = 0; i < topicPartitionCount; i++) {
+        partitionOffsetMap.put(Partition.of(i), cachedHeadOffsets.get(Partition.of(i)));
```

Comment:
You still need to do this asynchronously. The fanout means all of these lookups will be serialized. Use AsyncLoadingCache.java and Futures.allAsList instead to do this.

Reply:
I used https://github.com/ben-manes/caffeine, as the guava lib was soft deprecated (https://b.corp.google.com/issues/171496465, and many other places). I need to pull in some future conversion as well (similar to https://source.corp.google.com/search?q=net.javacrumbs.futureconverter.java8guava.FutureConverter.toCompletableFuture).